### PR TITLE
dialects: (stream) add constr to StreamType base class

### PR DESCRIPTION
**Pull request by superlopuh**
This is one of two cases of base classes of attributes that are used in constraints, but don't have a param_def.

### CODE REVIEW

File: xdsl/dialects/stream.py [hunk 1]
```diff
@@ -49,37 +47,93 @@ def __init__(self, element_type: _StreamTypeElement):
     def get_element_type(self) -> _StreamTypeElement:
         return self.element_type
 
+    @overload
+    @classmethod
```
Comment (alexarice) [inline xdsl/dialects/stream.py:51]: Having these as `@staticmethod`s should allow you to specify the generic type properly and avoid having to add `[Attribute]` below. Think a similar thing came up on the previous `constr` additions

Comment (superlopuh) [reply xdsl/dialects/stream.py]: Yep I had the same thought just pushed. Forgot that that was the conclusion of the previous iteration's discussion.

**Review by alexarice — APPROVED** [review]
Wonderful

File: xdsl/dialects/stream.py (hunk 1 above)
Comment (alexarice) [reply xdsl/dialects/stream.py]: I might go through and fix this in other places

Comment (alexarice) [reply xdsl/dialects/stream.py]: Unless that will clash with your next steps

Comment (alexarice) [reply xdsl/dialects/stream.py]: The only other one is in stencil and it looks like it'd be a real pain to get rid of

**Pull request merged**